user_group_settings: Disable dropdown if no permission to update it.

roanster007 opened the pull request:
Previously, even when a user doesn't have permission to edit a user group, the dropdown for the group permission settings was not disabled, and is possible for a user to change the values.

This commit adds a property `is_disabled` to the `dropdown_widget`, which can be used to disable the dropdown in cases of no permission  to update the widget, and the same is reflected in the user group permission settings.

<!-- Describe your pull request here.-->

[CZO thread](https://chat.zulip.org/#narrow/stream/9-issues/topic/group.20permissions.20dropdown.20not.20disabled/near/1886217)



<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://github.com/user-attachments/assets/d36f2eb0-9f8c-4e49-a51d-c955950ee4dd)
![image](https://github.com/user-attachments/assets/06c74821-9157-4442-8f80-c5a5c639506d)
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

roanster007 commented:
@Vector73 can you review this.

roanster007 commented:
@zulipbot add "buddy review"

Vector73 commented:
How about we show a tooltip when the dropdown is disabled - `You do not have permission to edit this group.`?

roanster007 commented:
>How about we show a tooltip when the dropdown is disabled - You do not have permission to edit this group.?

Makes sense, however, other dropdowns don't seem to have a tooltip. Let me confirm it in the czo.

sahil839 commented:
I guess the existing dropdowns are disabled in JS. So, would be better to do same here for now since that same functions can also be used to live-update the UI.

roanster007 commented:
>How about we show a tooltip when the dropdown is disabled - You do not have permission to edit this group.?

Added tooltip.

>I guess the existing dropdowns are disabled in JS. So, would be better to do same here for now since that same functions can also be used to live-update the UI.

Updated.

@Vector73 this is ready for another review.

Vector73 commented on web/src/user_group_edit.js:
```diff
@@ -64,31 +64,38 @@ export function get_edit_container(group) {
     );
 }
 
-function update_add_members_elements(group) {
+function update_group_settings_add_members_elements(group) {
     if (!is_editing_group(group.id)) {
```
Should we call this `update_group_settings_and_add_members_elements`?

roanster007 replied:
sure, updated!

Vector73 commented:
LGTM.

roanster007 commented:
@zulipbot add "mentor review"

roanster007 commented:
@zulipbot remove "buddy review"

roanster007 commented:
@prakhar1144 could you have a look at this.

prakhar1144 reviewed (commented):
Functionally it looks good. I suggested some code restructuring.

roanster007 commented on web/src/user_group_edit.js:
```diff
@@ -64,31 +64,38 @@ export function get_edit_container(group) {
     const $add_members_container = $(".edit_members_for_user_group .add_members_container");
 
+    // We are concerend with the General tab for changing group permissions.
+    const $group_permission_settings = $("#group_permission_settings");
```
I thought like since this would be the case with all of the permission dropdown elements that would be added in future, I would get the container of these dropdowns as `$group_permission_settings` and then get the dropdown elements there using `$group_permission_settings.find(".dropdown-widget-button");`.

So in future if we add more dropdowns, we just need to iterate on `$permission_dropdown_element` and disable them.

prakhar1144 replied:
Sounds good.

roanster007 commented on web/src/user_group_edit.js:
```diff
@@ -98,12 +105,19 @@ function update_add_members_elements(group) {
 function show_membership_settings(group) {
     const $edit_container = get_edit_container(group);
-    update_add_members_elements(group);
+    update_group_settings_and_add_members_elements(group);
```
updated!

roanster007 commented:
@prakhar1144
>I think the right approach should be to have separate show_membership_settings and show_general_settings and call them as required -- eventually update_add_members_elements and update_general_settings_elements ?

I addressed this review.

>I think we should just select the specific dropdown widget because the title says "Group Permissions" so we expect more permission settings to get added there with time. So name it something like can_mention_group_container...

For this, I replied in the comment, could you check it out... else I'll update it.

prakhar1144 commented on web/src/user_group_edit.js:
```diff
@@ -101,6 +101,31 @@ function update_add_members_elements(group) {
+
+    // Otherwise, we adjust whether the widgets are disabled based on
+    // whether this user is authorized to change the group settings.
+    const $permission_dropdown_element = $group_permission_settings.find(".dropdown-widget-button");
```
```suggestion
    const $permission_dropdown_elements = $group_permission_settings.find(".dropdown-widget-button");
```

roanster007 replied:
done!

prakhar1144 commented on web/src/user_group_edit.js:
```diff
@@ -101,6 +101,31 @@ function update_add_members_elements(group) {
+
+    if (settings_data.can_edit_user_group(group.id)) {
+        $permission_dropdown_element.prop("disabled", false);
+        $permission_dropdown_element.closest(".input-group")._tippy?.destroy();
```
I think we can add the tippy tooltip on the button element itself. No need to apply it to `.input-group`?

Currently it results in the tooltip even if we are away from the actual dropdown element:
<details><summary>Image</summary>
<img src="https://github.com/user-attachments/assets/8cfe1e2e-0348-4bcb-9fa3-2d49f6e11aac">
</img>
</details>

roanster007 replied:
I don't think tippy instances can be assigned to a button, so I assigned them to input-group. I guess I'll try to put a container around the button.

roanster007 replied:
seems like tippy instance can't be assigned to a button. One option is maybe to use `dropdown_widget_wrapper.hbs` so that I can assign the tippy instance to the `{{widget_name}}_widget_wrapper` but then it won't have the label.

A temporary solution I guess is to just leave it until we add the `tip` box like in other organization settings, but not sure what the right solution is.

prakhar1144 replied:
> seems like tippy instance can't be assigned to a button.

The first example on [tippyjs homepage](https://atomiks.github.io/tippyjs/) is for a button, so that's not quite correct :)

Probably it is due to we're trying to do things in bulk i.e. on all dropdowns. See `initialize_disable_btn_hint_popover` and check what's happening with the jquery object we're passing.

sahil839 replied:
The tooltip will not appear if added on a disabled button, as they do not show hover effects due to being disabled. So, we have been following this approach of using a container div element to add the tooltip for such cases.

But yeah we need to make sure that the tooltip is not visible when hovering away from the dropdown button.

prakhar1144 reviewed (commented):
Couple of comments, otherwise LGTM.

roanster007 commented on web/templates/dropdown_widget_with_label.hbs:
```diff
@@ -3,5 +3,7 @@
     </label>
     <span class="prop-element hide" id="id_{{widget_name}}" data-setting-widget-type="dropdown-list-widget" {{#if value_type}}data-setting-value-type="{{value_type}}"{{/if}}></span>
-    {{> dropdown_widget}}
+    <div class="inline-block dropdown_widget_with_label_wrapper">
+        {{> dropdown_widget}}
+    </div>
```
I made this wrapper here, since this seemed to be the only approach so as to show the tooltip only on the dropdown button.

roanster007 commented:
> I think we can add the tippy tooltip on the button element itself. No need to apply it to .input-group?
Currently it results in the tooltip even if we are away from the actual dropdown element:

@prakhar1144 addressed the pending review.

timabbott commented:
Since this makes HTML changes to the base HTML for all dropdown widgets, I'd like it to get pretty careful review; @amanagr can you look?

amanagr commented:
Looks good to me. I tested the dropdowns with label.

sahil839 commented:
There is some issue here with some of the dropdowns like there is no space between "Who can unsubscribe" setting and message retention setting -
![image](https://github.com/user-attachments/assets/2178f2a8-5cac-4085-854a-58049007b438)

And also some extra space between label and dropdown button in organization settings -
![image](https://github.com/user-attachments/assets/76328dc4-9534-4078-bc4f-061afbec6892)

roanster007 commented on web/styles/app_components.css:
```diff
@@ -1145,6 +1145,11 @@ div.overlay {
     }
 }
 
+.dropdown_widget_with_label_wrapper {
+    margin-top: 0 !important;
+    margin-bottom: 0 !important;
+}
+
```
I added these, because, in some places like organization permissions described in https://github.com/zulip/zulip/pull/30850#issuecomment-2252672757, margins are added to divs, causing the the issue described there.

roanster007 commented on web/styles/settings.css:
```diff
@@ -2065,6 +2065,10 @@ $option_title_width: 180px;
             margin-top: 12px;
         }
+
+        & .dropdown_widget_with_label_wrapper {
+            height: 30px;
+        }
```
I adjusted height of the wrapper wherever the `dropdown-widget-button` height is changed.

roanster007 commented:
>There is some issue here with some of the dropdowns like there is no space between "Who can unsubscribe" setting and message retention setting -

Had to make some changes to CSS and addressed the issue:
<details><summary>Updated Screenshots</summary>

![image](https://github.com/user-attachments/assets/891d5e22-9327-4aa5-bf0c-73b1d04e5175)
![image](https://github.com/user-attachments/assets/4ce974fe-cf26-4913-baf3-b177609d0b97)

</details>

sahil839 commented on web/styles/app_components.css:
```diff
@@ -1145,6 +1145,10 @@ div.overlay {
 }
 
+.dropdown_widget_with_label_wrapper {
+    margin-top: 0 !important;
```
We try to avoid overriding CSS properties using `important`. Maybe we can refactor the existing CSS which adds the 10px margin in some way. I guess the margin of 10px was added to have some space between the heading of first subsection and the setting label, so maybe we can refactor the CSS to do something like add a margin only on direct children `div` elements or `div` with a particular class.

timabbott replied:
div with a particular class is recommended; direct children selectors as the last selector in a rule cause performance issues for the rest of the app and are being purged.

timabbott commented:
Merged, thanks @roanster007 and reviewers!
